### ISSUE TITLE
Is there a way to get branch name in a user-defined rule?

### ISSUE THREAD

**Issue by and-semakin** [jorisroovers/gitlint#108]
By convention in our company we should put Jira ticket ID in the beginning of branch name (for example `JIRA-123_do_something_great`) and also we should put ticket id in commit message. I would like to check if they are the same. For this I need to get branch name inside of `validate` method of the rule. But I don't see that there is some property in `commit` or `context` objects that would help me. Do I miss something or gitlint is not able to help me in this case?

**Comment by jorisroovers** [comment]
So gitlint doesn't provide a way to do this out of the box, but I might have a workaround using the non-public `gitlint.git._git` function.

```python
from gitlint.rules import CommitRule, RuleViolation
from gitlint.options import IntOption
from gitlint.git import _git

class CheckBranchRule(CommitRule):
    name = "branch-name-match-title"
    id = "UC1"

    def validate(self, commit):
        print("title", commit.message.title)

        # See https://stackoverflow.com/a/12142066/381010
        print _git("rev-parse",  "--abbrev-ref", "HEAD")

        # I'll leave the comparison logic to you :-)

        return [RuleViolation(self.id, "Title and branch name don't match", line_nr=1)]
```

Caveats:
-  No guarantees about the availability or workings of the `gitlint.git._git` function in future gitlint releases, although I think it will remain fairly stable.
-  The code above will only work if you're trying to check the latest commit on the branch, it won't work when using gitlint's `--commits` argument to check multiple commits. This is because a commit might be part of multiple branches, so given a single sha it's not possible to figure out the *one* branch it's part of. At least not generically - in most cases the commit will only be part of a single branch of course. If you need this, you should still be able to get it to work using `git branch --contains commit.sha` and some extra parsing and business logic that fits your use-case.

Hope this helps!

**Comment by and-semakin** [comment]
Thanks for the quick reply @jorisroovers!
Looks reasonable. Will try it :)

Do you have plans to make branch name (or names) accessible via `commit` or `context` objects? For example it may be `List[str]` which will contain one or multiple branch names which commit belongs to.

**Comment by jorisroovers** [comment]
That did occur to me when writing the reply. 
Definitely not opposed to it, it's just a matter of finding/making the time. PRs help - a lot ;-)

**Comment by jorisroovers** [comment]
Good find, thanks for submitting the bug report :-)

There's 2 contributing factors here:

1. The integration between pre-commit and gitlint. When using pre-commit, gitlint does not fetch the commit message itself from the git repo, instead relying on pre-commit to pass the commit message via a file, using gitlint's ```--commit-msg``` flag.
2. When a commit message is passed via ```--commit-msg```, gitlint will not be querying the local git repository for additional information about the commit. This is because the last commit in the local repo might be a different commit from the one that is being linted via ```--commit-msg```.

This behavior will be the same when using gitlint's own commit-msg hook (```gitlint install-hook```) as it uses a similar mechanism under the hook (it passes the commit msg via stdin which causes gitlint to ignore local repo). I just verified that this is indeed the case.

What's really going on here is that gitlint currently doesn't really understand the concept of a pre-commit, instead relying on commit message passing to handle the pre-commit case. This is necessary as gitlint cannot read a non-committed message from the local repo.

I think the fix includes adding a ```--staged``` flag to tell gitlint to determine the set of staged files from the local repo and then use ```git diff --staged --name-only``` internally to determine the set of staged files.

Integrating this with pre-commit should just be a quick update to our [.pre-commit-hooks.yaml](https://github.com/jorisroovers/gitlint/blob/01addcdfc33f35995266ed75055dd361643c1e9d/.pre-commit-hooks.yaml) file.

I'll probably be able to work on this before the end of the year.

Meanwhile, you should be able to use this workaround with a user-defined rule:

```python
from gitlint.rules import CommitRule, RuleViolation
from gitlint.options import IntOption
from gitlint.git import _git

class CheckStagedFiles(CommitRule):
    name = "check-staged-files"
    id = "UC1"

    def validate(self, commit):
        print("body", commit.message.body) # string[]

        # NOTE: _git() is an internal, undocumented function. No guarantees on future availability or behavior :-)
        staged_files = _git("diff",  "--staged", "--name-only")

        # I'll leave the comparison logic to you :-)

        return [RuleViolation(self.id, "Staged file not mentioned in commit body", line_nr=1)]
```

Hope this helps!

**Comment by jorisroovers** [comment]
Implemented. Will release this as part of the 0.13.0 release, hopefully soon :-)

Also, I think it would be nice to also have a ```GitContext.current_branch``` property. Adding that to the plan for 0.14.0 (probably not any time soon).

**Comment by jorisroovers** [comment]
Released this yesterday as part of the 0.13.0 release - closing this out now :)
